wcfsetup/install/files/lib/action/FacebookAuthAction.class.php

```diff
@@ -31,125 +28,127 @@ class FacebookAuthAction extends AbstractAction {
+		$user = User::getUserByAuthData('facebook:'.$oauthUser->getId());
+		
+		if ($user->userID) {
+			if (WCF::getUser()->userID) {
+				// This account belongs to an existing user, but we are already logged in.
+				// This can't be handled.
+				throw new NamedUserException(
+					WCF::getLanguage()->getDynamicVariable('wcf.user.3rdparty.facebook.connect.error.inuse')
+				);
 			}
-			catch (SystemException $e) {
-				\wcf\functions\exception\logThrowable($e);
-				throw new IllegalLinkException();
+			else {
+				// This account belongs to an existing user, we are not logged in.
+				// Perform the login.
+				WCF::getSession()->changeUser($user);
+				WCF::getSession()->update();
+				HeaderUtil::redirect(LinkHandler::getInstance()->getLink());
 			}
+		}
+		else {
+			WCF::getSession()->register('__3rdPartyProvider', 'facebook');
 			
-			$userData = JSON::decode($content);
-			
-			// check whether a user is connected to this facebook account
-			$user = User::getUserByAuthData('facebook:'.$userData['id']);
-			
-			if ($user->userID) {
-				// a user is already connected, but we are logged in, break
-				if (WCF::getUser()->userID) {
-					throw new NamedUserException(WCF::getLanguage()->getDynamicVariable('wcf.user.3rdparty.facebook.connect.error.inuse'));
-				}
-				// perform login
-				else {
-					WCF::getSession()->changeUser($user);
-					WCF::getSession()->update();
-					HeaderUtil::redirect(LinkHandler::getInstance()->getLink());
-				}
+			if (WCF::getUser()->userID) {
+				// This account does not belong to anyone and we are already logged in.
+				// Thus we want to connect this account.
+				WCF::getSession()->register('__oauthUser', $oauthUser);
+
+				HeaderUtil::redirect(LinkHandler::getInstance()->getLink('AccountManagement').'#3rdParty');
 			}
 			else {
-				WCF::getSession()->register('__3rdPartyProvider', 'facebook');
-				// save data for connection
-				if (WCF::getUser()->userID) {
-					WCF::getSession()->register('__facebookUsername', $userData['name']);
-					WCF::getSession()->register('__facebookData', $userData);
-					
-					HeaderUtil::redirect(LinkHandler::getInstance()->getLink('AccountManagement').'#3rdParty');
-				}
-				// save data and redirect to registration
-				else {
-					WCF::getSession()->register('__username', $userData['name']);
-					if (isset($userData['email'])) WCF::getSession()->register('__email', $userData['email']);
-					WCF::getSession()->register('__facebookData', $userData);
-					
-					// we assume that bots won't register on facebook first
-					// thus no need for a captcha
-					if (REGISTER_USE_CAPTCHA) {
-						WCF::getSession()->register('noRegistrationCaptcha', true);
-					}
-					
-					WCF::getSession()->update();
-					HeaderUtil::redirect(LinkHandler::getInstance()->getLink('Register'));
-				}
+				// This account does not belong to anyone and we are not logged in.
+				// Thus we want to connect this account to a newly registered user.
+				
```

Comment:
Superfluous newline

Reply:
I instead opted to add the newlines everywhere, because they do not comment the statements directly below, but instead the "case" we are handling.